### PR TITLE
Java tests path issues

### PR DESCRIPTION
**Pull request by dmoody256**
Java tests on windows were broken when the default java paths where added in this commit: https://github.com/SCons/scons/commit/f48da72a3156d90bd6d58a840df1be2b013c5e9b

This was because of spacing issues when running some of the tests. This PR updates the test frame work to use quotes in all cases that the test is trying to run java found with the test environment and also takes care of removing quotes when needed.

This also includes a fix for the JARCHDIR test which was working around an issue with the FS.py function get_path.

Also other improvements to get the Java tests to reliably pass on windows environments including os.sep handling and regex improvements.

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: src/engine/SCons/Node/FS.py [hunk 1]
```diff
@@ -753,17 +753,27 @@ def srcnode(self):
...
         if self == dir:
             return '.'
         path_elems = self.get_path_elements()
         pathname = ''
         try: i = path_elems.index(dir)
         except ValueError:
-            for p in path_elems[:-1]:
-                pathname += p.dirname
+            # if dir arg was not set, then return the path from the base object
+            if dir_not_set:
```
Comment (bdbaddog) [inline src/engine/SCons/Node/FS.py:771]: Could you check if "dir is None" instead of having the flag?

**Comment by bdbaddog** [comment]
So the changes uses the location found by the test infrastructure, but what happens when you run SCons without specifying the path to the javac and java's installed in the default location?

By changing the tests are we no longer testing this?

**Comment by dmoody256** [comment]
> So the changes uses the location found by the test infrastructure, but what happens when you run SCons without specifying the path to the javac and java's installed in the default location?

It would work however it did before. I am only changing the test framework (well, besides the FS.py change, but that is not related to adding tools to the SCons environment path). 

I believe the way SCons works for java on linux is it picks up the java simlink in the /usr/bin. I would guess that one of the other common linux tools adds this path to the SCons environment, and so the tests on Java have always worked because of that. They never worked on windows, and would get skipped because there were no default paths before the commit listed in the PR description. Then after default paths were added, the windows tests started failing because of unquoted spaces in the java commands.

> By changing the tests are we no longer testing this?

Well, what I think you're suggesting (default java paths in SCons environment) is not a feature of SCons to test. It just happened to work on linux, because the /usr/bin ends of in the default SCons environment, which is where the java binaries usually get simlinked on linux. So by specifying the paths in all the test, yes, we would not be testing Java being in the PATH in those environments.

**Comment by bdbaddog** [comment]
Seems like if it's failing now (on Windows) because the path SCons determines has a space in it which is not being escaped, then that's the correct behavior, and the more correct fix would be to fix the escaping?

**Comment by dmoody256** [comment]
> Seems like if it's failing now (on Windows) because the path SCons determines has a space in it which is not being escaped, then that's the correct behavior, and the more correct fix would be to fix the escaping?

Yeah, that's what this PR is, it doesn't change what the tests are testing, but makes them more robust in different environments. 

An improvement to SCons would be adding default path logic to the java tools so that they can find standard java installations automatically. Similar to the logic that is already in the test framework,

**Comment by bdbaddog** [comment]
I'd rather not fix the tests to work when SCons wouldn't if the tool paths are reasonable defaults.
Can we just fix the noted commit?
Would escaping the paths there help?

**Comment by dmoody256** [comment]
The problem is not at java_ENV where the test framework sets up default java paths. Those should be un-escaped because python handles the escaping and the java binaries get appended to the unescaped path. 

If you don't want to change the test then the scons action may need to do a better job of handling the COM strings when there are spaces. It seems that this findall in substitute splits up the commands with paths which include spaces:
https://github.com/SCons/scons/blob/90b285240e9d80a6f4b2eb9d294ba9e789fd269d/src/engine/SCons/Subst.py#L732-L742

I noticed when I tested that substitute function with the quoted java commands from my PR, it was still separating the java binary path (ex: '"C:\Program Files...\javac.EXE"' would substituted like this ['"C:\Programs', 'Files...\javac.EXE"']), but it would end up working when it got back to the command line. Python thought it was passing two args, then windows would interpret it as one because of the quotes.

I haven't reduced this into a smaller generic test and I think it works with other tools so I am not sure at this point.

**Comment by bdbaddog** [comment]
exactly.. fixing the escaping in the actions is the right path.
Thanks for your persistance!

**Comment by dmoody256** [comment]
> exactly.. fixing the escaping in the actions is the right path.

I was looking into it and I don't think it can be done in the Action class. It is too generic at that level, because really any kind of string could be passed through the Action and there would not be a way to tell what are the args and what is a space separated command.

I did find that I could do some modifications at the Tool level since there I know what part of the command is going to be the binary and what part is going to be the args. I came up with something like this for the builders in ```Tool/__init__.py```:
```python
def binary_file_wrapper(bin_key):

    def bin_file(target, source, env):
        if ' ' in env[bin_key]:
            env[bin_key] = env.File(env[bin_key])
    def emptyStrfunction(self, target, source, env, executor=None):
            return ''

    bin_file_action = SCons.Action.FunctionAction( bin_file, {} )
    bin_file_action.strfunction = emptyStrfunction
    return bin_file_action
```

Then down in the builders, I use this for the JAR, or JAVAC command and make an action list which makes sure the binary is escaped by making it a file:
```python
def CreateJarBuilder(env):
#...
#...
        jar_com = SCons.Action.Action('$JARCOM', '$JARCOMSTR')
        java_jar = SCons.Builder.Builder(action = SCons.Action.ListAction([
                                            binary_file_wrapper('JAR'), 
                                            jar_com]),
                                         suffix = '$JARSUFFIX',
                                         src_suffix = '$JAVACLASSSUFFIX',
                                         src_builder = 'JavaClassFile',
                                         source_factory = fs.Entry)
```

Thoughts on this approach? This could be used for any other builder, not just Java builders.

**Comment by dmoody256** [comment]
> I'm suspicious this is a quoting problem in the tests due to the fact that the test SConstruct is written as a string to a file. switching to file fixtures might be the answer

So we can fix the tests or fix the engine to handle it. Using a quote or File object in the tests should work, but bdbaddog was suggesting we fix the code to handle it so the user (test) doesn't have to. Also the keeping the test the same maintains the original intent of the engine code, even if it never worked.

The current code in review changes the tests, I am still waiting for some feedback regarding the changes to the Tools I proposed in the [comment](https://github.com/SCons/scons/pull/3176#issuecomment-414475728) above.

**Comment by mwichmann** [comment]
So a little more experimenting here. scons indeed doesn't deal right with a tool name set in the environment that has an embedded space.  I ran the CC tests which pass without complaint.  I then poked in the actual path to the c compiler, and it fails.  So the SConstruct for the CC test at this point was:
```
foo = Environment()
foo['CC'] = r'c:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\cl.exe'
print("=== C Compiler is: ===")
print(foo.Dump('CC'))
cc = foo.Dictionary('CC')
bar = Environment(CC = r'c:\Users\mats\Envs\PY37\Scripts\python.exe wrapper.py ' + cc)
print("=== Wrapped C Compiler is: ===")
print(foo.Dump('CC'))
foo.Program(target = 'foo', source = 'foo.c')
bar.Program(target = 'bar', source = 'bar.c')
```
leading to:
```
scons: Reading SConscript files ...
=== C Compiler is: ===
'c:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\cl.exe'
=== Wrapped C Compiler is: ===
('c:\\Users\\mats\\Envs\\PY37\\Scripts\\python.exe wrapper.py c:\\Program '
 'Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\cl.exe')
scons: done reading SConscript files.
scons: Building targets ...
c:\Users\mats\Envs\PY37\Scripts\python.exe wrapper.py c:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\cl.exe /Fobar.obj /c bar.c /nologo
'c:\Program' is not recognized as an internal or external command,
operable program or batch file.
link /nologo /OUT:bar.exe bar.obj
LINK : fatal error LNK1181: cannot open input file 'bar.obj'
scons: *** [bar.exe] Error 1181
scons: building terminated because of errors.
```
I'm left with questions... I guess we don't have a test that forces an embedded space into to a tool where it matters - as noted above there's a reason to deal with those strings as possibly being "multiple words" - as in fact the step above with the wrapper tries to test - so something needs more careful handling.  The suggestion above looks reasonable, although it would be even better if it could be "invisible".  Note in this case scons decided to build the "bar.obj" target first.

Aside: so why does Java tests need special-cased routines in to determine the path, that are different from the ones scons itself uses? Seems like none of the other tests do this. Running scons itself, JAVAC and JAR variables remain set to simple names (without the paths), and stuff seems to work.

**Review by mwichmann — COMMENTED** [review]
I've been mucking around in the Java tests just now and generally would like to see these changes go in.  In particular, adding the settings in the Environment call; plus the os.sep addition in JAVABOOTCLASSPATH and the fixup in JAVACLASSPATH and the nested-classes change, the latter three of which are all errors causing fails that have nothing to do with the irritating quoting problem for the java executable itself.  I'm reluctant to submit my changes at this point as I think a number of them will cause conflicts that need resolution, but I will say that I'm going to propose changing the way java_where_javac operates - it's got several bugs (which don't allow it to return a version number in some cases, which breaks the RMIC tests here and in Repository) and really doesn't need the complexity of invoking javac -version through the run method chain - a simple subprocess.check_output call will work fine for that.  If Bill is reluctant to proceed, maybe we can unentangle most of this from the command-name-quoting-for-Windows problem and do that part separately?

**Comment by bdbaddog** [comment]
Declining this pull request.
I've reimplemented such that the javac, etc tools don't need to be passed around in quotes.
See #3206

**Pull request closed without merging**